File: lc/DFS/200. Number of Islands.go
func numIslands(grid [][]byte) int {
    
    res := 0

    // 这里要先定义一个函数签名，要不然函数里面无法调用自己
    var dfs func(i, j int) bool
    dfs = func (i, j int) bool {
        if i < 0 || i >= len(grid) || j < 0 || j >= len(grid[0]) || string(grid[i][j]) == "0" {
            return false
        }

        grid[i][j] = []byte("0")[0]

        dfs(i+1, j)
        dfs(i-1, j)
        dfs(i, j+1)
        dfs(i, j-1)

        return true
    }

    for i:=0; i<len(grid); i++ {
        for j:=0; j<len(grid[0]); j++ {
            if dfs(i,j) {
                res += 1
            }
        }
    }

    return res
}

File: lc/DFS/200. Number of Islands.ts
function numIslands(grid: string[][]): number {
    let res = 0;

    const dfs = (i: number, j: number): boolean => {
        if (i < 0 || i >= grid.length || j < 0 || j >= grid[0].length || grid[i][j] === "0") {
            return false;
        }
        grid[i][j] = "0";
        dfs(i+1, j)
        dfs(i-1, j)
        dfs(i, j+1)
        dfs(i, j-1)

        return true
    }

    for (let i = 0; i < grid.length; i++) {
        for (let j=0; j<grid[0].length; j++) {
            if (grid[i][j] === "1") {
                res += 1
                dfs(i,j)
            }
        }
    }

    return res
};